dashboard/app/models/script_level.rb

```diff
@@ -372,6 +372,58 @@ def summarize_as_bonus
     }.camelize_keys
   end
 
+  # Bring together all the information needed to show the teacher panel on a level
+  def summarize_for_teacher_panel(student)
+    contained_levels = levels.map(&:contained_levels).flatten
+    if contained_levels.any?
+      user_level = student.last_attempt_for_any(contained_levels)
+      contained = true
+    else
+      user_level = student.last_attempt_for_any(levels)
+      contained = false
+    end
+
+    status = activity_css_class(user_level)
+    passed = [SharedConstants::LEVEL_STATUS.passed, SharedConstants::LEVEL_STATUS.perfect].include?(status)
+
+    if user_level
+      paired = user_level.paired?
+
+      driver_info = if contained
+                      UserLevel.most_recent_driver(script, contained_levels, student)
+                    else
+                      UserLevel.most_recent_driver(script, level, student)
```

Comment:
in terms of style, this could read easier as 
```
levels = contained ? contained_levels : [level]
driver_info = UserLevel.most_recent_driver(script, levels, student)
...
navigator_info = UserLevel.most_recent_navigator(script, levels, student)
```
could also use this to eliminate `if contained_levels.any?` above.

you could omit the `[]` around `[level]`... i just added them to make the name "levels" more accurate 😬